images/ami_kms_key.tf

```diff
@@ -51,20 +51,34 @@ data "aws_iam_policy_document" "ami_kms_doc" {
   statement {
     sid = "Allow use of the key"
 
+    # Wildcards (other than the global "*") are not allowed when specifying
+    # a principal (e.g. "${aws_iam_role.ec2amicreate_role.arn}*"), so instead
+    # we set the principal to "*" and restrict access via the condition
+    # below (which does allow for a wildcard pattern match on the role ARN)
     principals {
       type        = "AWS"
-      identifiers = ["${aws_iam_role.ec2amicreate_role.arn}"]
+      identifiers = ["*"]
     }
 
     actions = [
+      "kms:CreateGrant",
       "kms:Decrypt",
       "kms:DescribeKey",
       "kms:Encrypt",
       "kms:GenerateDataKey*",
       "kms:ReEncrypt*",
+      "kms:RetireGrant",
     ]
 
     resources = ["*"]
+
+    condition {
+      test     = "StringLike"
+      variable = "aws:PrincipalArn"
+      values = [
+        "${aws_iam_role.ec2amicreate_role.arn}*"
```

Comment:
Nice trick!

Reply:
I can't take credit for it- I was trying to use `test = "ArnLike"` and it wasn't working (because I didn't read the docs carefully enough), but then I stumbled across [this helpful nugget](https://stackoverflow.com/a/56145538).

Reply:
🎼  Yeah, yeah, oh-yeah, what **condition** my **condition** was in 🎵